core/src/validator.rs

```diff
@@ -24,234 +20,236 @@ pub enum Error {
-                         when removing a validator. This is a bug",
-                )
-                .retain(|validator_id| validator_id != id);
-            true
+    /// Fails if failed to load wasm blob.
+    pub fn new(raw_validator: data_model_validator::Validator) -> Result<Self> {
+        let engine = wasm::create_engine();
+        Ok(Self {
+            loaded_validator: LoadedValidator::load(&engine, raw_validator)?,
+            engine,
         })
     }
 
-    /// Validate given `operation` with all [`Chain`] validators of required type.
-    ///
-    /// If no validator with required type is found, then return [`Ok`].
+    /// Validate operation.
     ///
     /// # Errors
     ///
-    /// Will abort the validation at first
-    /// [`Deny`](iroha_data_model::permission::validator::Verdict::Deny) validator verdict and
-    /// return an [`Err`](Result::Err).
-    ///
-    // TODO: Possibly we can use a separate validator thread
-    #[allow(clippy::expect_used, clippy::unwrap_in_result)]
+    /// - Failed to prepare runtime for WASM execution;
+    /// - Failed to execute WASM blob;
+    /// - Validator denied the operation
     pub fn validate(
         &self,
         wsv: &WorldStateView,
         authority: &<Account as Identifiable>::Id,
-        operation: impl Into<NeedsPermissionBox>,
+        operation: impl Into<data_model_validator::NeedsPermissionBox>,
     ) -> Result<()> {
         let operation = operation.into();
-        let Some(validators) = self
-            .concrete_type_validators
-            .get(&operation.required_validator_type()) else
-        {
-            return Ok(())
-        };
 
         let runtime = wasm::RuntimeBuilder::new()
             .with_engine(self.engine.clone()) // Cloning engine is cheap, see [`wasmtime::Engine`] docs
             .with_configuration(wsv.config.wasm_runtime_config)
             .build()?;
 
-        for validator_id in validators.value() {
-            self.execute_validator(&runtime, wsv, authority, validator_id, &operation)?
-        }
+        trace!("Running validator");
+        let verdict = runtime.execute_permission_validator_module(
+            wsv,
+            authority,
+            &self.loaded_validator.module,
+            &operation,
+        )?;
 
-        Ok(())
+        Result::<(), data_model_validator::DenialReason>::from(verdict).map_err(|reason| {
+            Error::ValidatorDeny {
+                operation: operation.clone(),
+                reason,
+            }
+        })
     }
+}
 
-    /// Get constant view to the [`Chain`] without interior mutability
-    pub fn view(&self) -> ChainView {
-        ChainView { chain: self }
+/// Mock of validator for unit tests of `iroha_core`.
+///
+/// We can't use real validator because WASM for it is produced in runtime from outside world.
+#[cfg(test)]
+#[derive(Default, Debug, Copy, Clone)]
+pub struct MockValidator;
```

Comment:
```suggestion
#[cfg(test)]
pub struct MockValidator;
```

you should strive to make it look as the original validator

Reply:
Actually all this trait impls are used in `wsv` except for `Copy` wich is needed because of our lints, but I have no objections about `Copy`